libp2p/protocols/secure/noise.nim

```diff
@@ -427,10 +437,24 @@ method write*(sconn: NoiseConnection, message: seq[byte]): Future[void] {.async.
   while left > 0:
     let
       chunkSize = min(MaxPlainSize, left)
-      cipher = sconn.writeCs.encryptWithAd(
-        [], message.toOpenArray(offset, offset + chunkSize - 1))
 
-    await sconn.stream.writeFrame(cipher)
+    var
+      cipherFrame =
+        newSeqUninitialized[byte](2 + chunkSize + sizeof(ChaChaPolyTag))
+
+    # Frame consists of length + cipher data + tag
+    cipherFrame[0..<2] = toBytesBE(uint16(chunkSize + sizeof(ChaChaPolyTag)))
+
+    copyMem(addr cipherFrame[2], unsafeAddr message[offset], chunkSize)
```

Comment:
What's the purpose of the `offset` variable here? Any value besides 0 would be unsafe in this `copyMem` call.

Reply:
`offset` marks the start of the chunk within `message` - it's increased by `chunkSize` on every loop iteration and `chunkSize` is capped by `message.len` - why is it unsafe?

I'm slightly uncomfortable with `copyMem` here, but unfortunately, the std lib does an extremely slow byte-by-byte iteration here that turns up in benchmarking

Reply:
Apologies, I've failed to notice some of the hidden lines in the diff.